sched/environ/env_setenv.c

```diff
@@ -144,36 +144,42 @@ int setenv(FAR const char *name, FAR const char *value, int overwrite)
 
   if (group->tg_envp)
     {
-      newsize = group->tg_envsize + varlen;
-      newenvp = (FAR char *)group_realloc(group, group->tg_envp,
-                                          newsize + 1);
+      while (group->tg_envp[oldsize] != NULL)
+        {
+          oldsize++;
+        }
+
+      newenvp = group_realloc(group, group->tg_envp,
+                              sizeof(*newenvp) * (oldsize + 2));
       if (!newenvp)
         {
           ret = ENOMEM;
           goto errout_with_lock;
         }
 
-      pvar = &newenvp[group->tg_envsize];
+      newenvp[oldsize + 1] = NULL;
     }
   else
     {
-      newsize = varlen;
-      newenvp = (FAR char *)group_malloc(group, varlen + 1);
+      oldsize = 0;
+      newenvp = group_zalloc(group, sizeof(*newenvp) * 2);
       if (!newenvp)
         {
           ret = ENOMEM;
           goto errout_with_lock;
         }
-
-      pvar = newenvp;
     }
 
-  newenvp[newsize] = '\0';
+  group->tg_envp = newenvp;
 
-  /* Save the new buffer and size */
+  pvar = group_malloc(group, varlen);
```

Comment:
Maybe we can do allocation of `pvar` before realloc/zalloc above just to have better error handling?

Reply:
Actually, the initial code just like what you suggest above, but I change it to the current form, because we don't need release newenvp or pvar at all.

Reply:
I mean that if `pvar` allocation fails the code
```
      while (group->tg_envp[envc] != NULL)
        {
          envc++;
        }
```
will con provide us a reliable information about length of memory pointed by `group->tg_envp` during the next API call. I mean that if allocation of `pvar` fails early then we can bypass all the reallocation code and return error on early stages. If that makes sense to you

Reply:
Ok, done.